### PR TITLE
Search API design and documentation [SA-8836]

### PR DESCRIPTION
**Pull request by armatronic**
I've designed an API which should more-or-less fit the most recent filter and output layouts.

**Notes**
- badges, events, news, pages and users are in the first level of search results
- pages contain their own search results for assessments, files, forums, links, text boxes and tiles

**TODO**
- [x] specify required fields
- [x] examples
- [x] fix build warnings?
  - ~~I haven't been able to fix them all, but I don't know if Redocly supports everything I want to do with discriminators~~ `npm update` was enough

### CODE REVIEW

**Comment by jleckie** [comment]
This looks pretty good to me @armatronic.  It seems to cover all the filter options, and return type objects.

File: openapi/paths/api@search@folder@{id}.yaml [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  responses:
+    '200':
+      $ref: ../components/responses/search/pageSubResult-list.yaml
+    default:
+      $ref: ../components/responses/problem.yaml
+  parameters:
+    - $ref: ../components/parameters/id.yaml
+    - $ref: ../components/parameters/search/keyword.yaml
+    - $ref: ../components/parameters/search/filter.yaml
+    - $ref: ../components/parameters/cursor.yaml
```
Comment (armatronic) [inline openapi/paths/api@search@folder@{id}.yaml:17]: Note to self: add `limit` query string parameter here

**Comment by tallzeebaa** [comment]
> fix build warnings?

Yes, please... It's worth looking at them.

File: openapi/components/examples/search/result.yaml [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+        - id: 3
+          name: Year 1
+          url: https://{hostname}.{domain}/homepage/3
+      icon: homepage
+      _links:
+        subResults: https://{hostname}.{domain}/api/search/folder/10
+      _subResults:
+        data:
+          - id: 1
+            type: assessment.dueWork
```
Comment (MichaelHeydon) [inline openapi/components/examples/search/result.yaml:81]: This is definitely an _Assessment_ and not a _Learning Activity_?

File: openapi/components/schemas/search/resultTypes/assessment.yaml [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+          - assessment.project: this is a project
+            (activities within the project are not fetched)
+          - assessment.task: this is a task
+          - assessment.quiz: this is a quiz
+        enum:
+          - assessment.dueWork
+          - assessment.lessonPlan
+          - assessment.project
+          - assessment.task
+          - assessment.quiz
```
Comment (MichaelHeydon) [inline openapi/components/schemas/search/resultTypes/assessment.yaml:22]: Would `ltiAssessment` also be in here?

File: openapi/components/examples/search/result.yaml (hunk 2 above)
Comment (armatronic) [reply openapi/components/examples/search/result.yaml:81]: I am not sure :crying_cat_face:

It ought to be able to be changed fairly easily?

File: openapi/components/schemas/search/resultTypes/assessment.yaml (hunk 3 above)
Comment (armatronic) [reply openapi/components/schemas/search/resultTypes/assessment.yaml:22]: Apparently they aren't intended to be searchable, but I'd not have an objection to adding them.

**Pull request merged**